File: 23_merge_k_sorted_lists/leetcode_23_priorityqueue.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

from queue import PriorityQueue

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:

        head = cur = ListNode(-1)
        q = PriorityQueue()

        for l in lists:
            if l:
                q.put((l.val, l))

        while not q.empty():
            val, node = q.get()
            if node.next:
                q.put((node.next.val, node.next))
            cur.next = node
            cur = cur.next

        return head.next

File: 23_merge_k_sorted_lists/leetcode_23_priorityqueue.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        // edge case []
        if (lists.length == 0) return null;
        
        ListNode dummy = new ListNode(-1), cur=dummy;

        PriorityQueue<ListNode> queue = new PriorityQueue<>(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode n1, ListNode n2) {
                return n1.val-n2.val;
            }
        });

        for (ListNode head: lists) {
            if (head != null) queue.add(head);
        }

        while(!queue.isEmpty()) {
            ListNode node = queue.poll();
            if (node.next != null) queue.add(node.next);
            cur.next = node;
            cur = cur.next;
        }

        return dummy.next;
    }
}